### PR TITLE
fix defaults for core jobs

### PR DESCRIPTION
**Pull request by nickboldt**

### CODE REVIEW

File: buildchow/templates/aggregate-site-template.yaml [hunk 1]
```diff
@@ -10,11 +10,11 @@
 
     node: "{nodes}"
     logrotate:
-      daysToKeep: 5
```
Comment (maxandersen) [inline buildchow/templates/aggregate-site-template.yaml]: tip: if we are going to use the exact same logrotate for all (makes sense) then that could be done as a macro.

File: buildchow/templates/core-jobs.yaml [hunk 2]
```diff
@@ -5,9 +5,11 @@
       master
     devstudio-stream:
       master
-    target-platform-version: 4.50.0.Beta1-SNAPSHOT
-    target-platform-version-maximum: 4.50.0.Beta1-SNAPSHOT
-    disabled: true
+    target-platform-version: 4.50.0.CR1-SNAPSHOT
+    target-platform-version-maximum: 4.50.0.CR1-SNAPSHOT
+    blockBuildWhenDownstreamBuilding: false
```
Comment (maxandersen) [inline buildchow/templates/core-jobs.yaml]: interesting. Do we really want this ?

I'm not sure, but one advantage having this true is
that if you push two related changes (i.e. one in base and one in openshift) you don't see openshift fail because it is missing the change from base ?

File: buildchow/templates/core-jobs.yaml [hunk 3]
```diff
@@ -42,17 +44,17 @@
...
     stream:
       4.3.mars
     devstudio-stream:
       9.0.mars
-    target-platform-version: 4.50.0.Beta1
-    target-platform-version-maximum: 4.50.0.Beta1
-    branch: jbosstools-4.3.0.Beta1x
+    target-platform-version: 4.50.0.CR1
+    target-platform-version-maximum: 4.50.0.CR1
+    branch: jbosstools-4.3.x
```
Comment (maxandersen) [inline buildchow/templates/core-jobs.yaml]: this haven't been created yet so it would fail

File: buildchow/templates/defaults.yaml [hunk 4]
```diff
@@ -4,35 +4,21 @@
...
     default-remote: origin
-    jvm-memory: "-Xms512m -Xmx1024m -XX:PermSize=128m -XX:MaxPermSize=256m"
+    # since we use JDK 8 now, no need for these:  -XX:PermSize=128m -XX:MaxPermSize=256m"
+    jvm-memory: "-Xms512m -Xmx1536m"
     tests-folder: tests # for some it is all-tests
     component-responsible: ""
     scm-schedule: H H/8 * * * # every 8 hours starting a random hour. Removes need for each job to configure it (http://htmlpreview.github.io/?https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/hudson/triggers/TimerTrigger/help-spec.html)
-    download-cache:  "$JENKINS_HOME/static_build_env/jbds/download-cache"
-    maven-settings: "$JENKINS_HOME/../.m2/local-settings.xml" # need more reliable location
+    JENKINS_HOME: "/home/hudson"
```
Comment (maxandersen) [inline buildchow/templates/defaults.yaml]: why this change ? This means you are hardcoding it to /home/hudson and don't use the $JENKINS_HOME as actually defined by the jenkins you are running on.

Makes the templates non-portable.

File: buildchow/templates/product-template.yaml [hunk 5]
```diff
@@ -47,16 +45,15 @@
...
+          default: https://devstudio.redhat.com/9.0/snapshots/builds/jbosstools-build-sites.aggregate.central-site_${{jbosstools_site_stream}}/latest/all/repository.zip
       - bool:
           name: skipRevisionCheckWhenPublishing
           default: false
           description: |
             Check box to always publish new build to staging / nightly. <br/>
             Unchecked, this job will check previous build's revision to decide if this build needs <br/>
-            to overwrite the previous one, or if it's in fact the same source &amp; therefore same binaries.<br/>
-      
+            to overwrite the previous one, or if it's in fact the same source & therefore same binaries.
```
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: hmm - don't you need to still escape the & to make it show up properly ?

File: buildchow/templates/product-template.yaml [hunk 6]
```diff
@@ -105,7 +97,7 @@
                # now rebuild the update site, without pack200
       - maven-target:
           goals: "clean install ${{MAVEN_FLAGS}} -DTARGET_PLATFORM_VERSION=${{TARGET_PLATFORM_VERSION_MAXIMUM}}"
-          maven-version: maven-3.1.1
+          maven-version: maven-3.2.5
```
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: should this not just use `{default-maven-version}` ?

File: buildchow/templates/product-template.yaml [hunk 7]
```diff
@@ -117,39 +109,44 @@
...
             - download.cache.directory={download-cache}
             - targetPlatformGroup=jbdevstudio
-            - jbosstools-site=http://download.jboss.org/jbosstools/updates/nightly/core/${{jbosstools_site_stream}}/
+            - jbosstools-site=http://download.jboss.org/jbosstools/mars/snapshots/builds/jbosstools-build-sites.aggregate.site_${{jbosstools_site_stream}}/latest/all/repo/
           pom: ${{WORKSPACE}}/sources/product/site/pom.xml          
       - maven-target:
-          goals: "org.apache.maven.plugins:maven-dependency-plugin:2.9:unpack"
-          maven-version: maven-3.x
+          goals: "-U org.apache.maven.plugins:maven-dependency-plugin:2.9:unpack"
+          maven-version: maven-3.2.5
```
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: {default-maven-version}

File: buildchow/templates/product-template.yaml [hunk 8]
```diff
@@ -117,39 +109,44 @@
             - maven.test.error.ignore=true
             - download.cache.directory={download-cache}
             - targetPlatformGroup=jbdevstudio
-            - jbosstools-site=http://download.jboss.org/jbosstools/updates/nightly/core/${{jbosstools_site_stream}}/
+            - jbosstools-site=http://download.jboss.org/jbosstools/mars/snapshots/builds/jbosstools-build-sites.aggregate.site_${{jbosstools_site_stream}}/latest/all/repo/
           pom: ${{WORKSPACE}}/sources/product/site/pom.xml          
       - maven-target:
-          goals: "org.apache.maven.plugins:maven-dependency-plugin:2.9:unpack"
-          maven-version: maven-3.x
+          goals: "-U org.apache.maven.plugins:maven-dependency-plugin:2.9:unpack"
```
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: why the -U ? 

At what point is it necessary to check that 2.9 is actually still 2.9 ? Don't -U only make sense for builds dependent on snapshots ?

File: buildchow/templates/product-template.yaml [hunk 9]
```diff
@@ -117,39 +109,44 @@
...
+          maven-version: maven-3.2.5
           private-repository: false
           properties:
-            - artifact=org.jboss.tools.releng:jbosstools-releng-publish:4.3.0.Alpha1-SNAPSHOT:zip
+            - artifact=org.jboss.tools.releng:jbosstools-releng-publish:4.3.0.CR1-SNAPSHOT:zip
```
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: this seem to be referred multiple places - extract a value in `defaults.yaml` or in project definitions to reuse and ensure using the same consistently ?

File: buildchow/templates/product-template.yaml [hunk 10]
```diff
@@ -117,39 +109,44 @@
...
-               
-               cd ${{WORKSPACE}}/sources/
-               export skipRevisionCheckWhenPublishing=${{skipRevisionCheckWhenPublishing}}
-               DESTINATION=/qa/services/http/binaries/RHDS . publish/publish.sh
+
+               ###############################################
+
+               if [[ ${{skipRevisionCheckWhenPublishing}} == "true" ]] || [[ $(. ${{WORKSPACE}}/sources/util/checkLatestPublishedSHA.sh -s ${{WORKSPACE}}/sources/product/site/target/repository -t http://www.qa.jboss.com/binaries/RHDS/9.0/snapshots/builds/${{JOB_NAME}}/latest/all/repo/) == "true" ]]; then
+                 cd ${{WORKSPACE}}/sources/product/results
+                 M2_HOME=/qa/tools/opt/apache-maven-3.2.5/
```
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: this is making it non-portable. Why not use the environment variables that define where maven version is ?

File: buildchow/templates/product-template.yaml [hunk 11]
```diff
@@ -117,39 +109,44 @@
...
+                 BUILD_DESCRIPTION="NOT PUBLISHED: UNCHANGED"
+               fi
+
     publishers:
         - description-setter: 
             regexp: "BUILD_DESCRIPTION='(.+)'"
             description: "\\1"
-            ## TODO: downstreamtrigger
+        - raw:
+            xml: |
```
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: any reason to not use the native built-in support ? 

http://docs.openstack.org/infra/jenkins-job-builder/publishers.html#publishers.downstream-ext

File: buildchow/templates/product-template.yaml [hunk 12]
```diff
@@ -160,5 +157,6 @@
             publish-failed-builds: false
 
     wrappers: 
-        - timeout:
-            timeout: 120
+        - raw:
+            xml: |
+               <hudson.plugins.build__timeout.BuildTimeoutWrapper><strategy class="hudson.plugins.build_timeout.impl.AbsoluteTimeOutStrategy"><timeoutMinutes>180</timeoutMinutes></strategy><operationList><hudson.plugins.build__timeout.operations.AbortOperation/></operationList></hudson.plugins.build__timeout.BuildTimeoutWrapper>
```
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: any reason to not use the built-in support ? 

http://docs.openstack.org/infra/jenkins-job-builder/wrappers.html#wrappers.timeout

File: buildchow/templates/core-jobs.yaml (hunk 2 above)
Comment (nickboldt) [inline buildchow/templates/core-jobs.yaml]: It's already in the jobs, so yes. Also, since we removed inter-job linkages 2+ years ago, there's no way for Jenkins to know that base blocks openshift.

File: buildchow/templates/core-jobs.yaml (hunk 3 above)
Comment (nickboldt) [inline buildchow/templates/core-jobs.yaml]: Correct, but it needs to be in place for code freeze.

File: buildchow/templates/defaults.yaml (hunk 4 above)
Comment (nickboldt) [inline buildchow/templates/defaults.yaml]: can JENKINS_HOME be set in the buildchow/configs/default/jenkins_jobs.ini so it's a global var at the server level?

File: buildchow/templates/product-template.yaml (hunk 5 above)
Comment (nickboldt) [inline buildchow/templates/product-template.yaml]: Nope.

File: buildchow/templates/product-template.yaml (hunk 6 above)
Comment (nickboldt) [inline buildchow/templates/product-template.yaml]: Good idea.

File: buildchow/templates/product-template.yaml (hunk 8 above)
Comment (nickboldt) [inline buildchow/templates/product-template.yaml]: I understood the -U would impact the version of artifact downloaded, not the version of plugin run. Is that wrong?

File: buildchow/templates/core-jobs.yaml (hunk 2 above)
Comment (maxandersen) [inline buildchow/templates/core-jobs.yaml]: if there are no interlinkages why are they in the jobs ? What you just wrote seems contradictory.
If we don't have linkages then this value means nothing/has no effect so it is not necessary to add in, right ?

File: buildchow/templates/core-jobs.yaml (hunk 3 above)
Comment (maxandersen) [inline buildchow/templates/core-jobs.yaml]: why not use the right branch so we can actually test these running locally?

File: buildchow/templates/defaults.yaml (hunk 4 above)
Comment (maxandersen) [inline buildchow/templates/defaults.yaml]: not following. why do you need it in jenkins job level ? $JENKINS_HOME is defined in jenkins already. JJB is better of *not* knowing the fixed path.

File: buildchow/templates/product-template.yaml (hunk 8 above)
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: ah, yeah - I guess if we continue to use releng snapshot then it is necessary.

But:

a) shouldn't releng be released before codefreeze so we stop using snapshots for our builds ? (that is one of the reasons we wanted to get away from just using latest master version of releng scripts)

b) higher up we have:
       `goals: "{ci-maven-flags} org.apache.maven.plugins:maven-dependency-plugin:2.9:unpack" #TODO: should it have all MAVEN_FLAGS ?`
  so it does the same thing; but here it is usign ci-maven-flags. They should at least use the same approach.

Comment (nickboldt) [inline buildchow/templates/product-template.yaml]: Re: b: Good point.

Re: a: yes, in theory. Talk to Mickael about how/when we should release. I'm good using snapshots and just branching with every milestone. I don't see the value-add for actually doing nexus releases.

File: buildchow/templates/core-jobs.yaml (hunk 2 above)
Comment (nickboldt) [inline buildchow/templates/core-jobs.yaml]: Some jobs are linked (aggregates, JBDS installer, tests). Most are not (jbosstools-* component jobs). So it helps the linked jobs and doesn't do anything for the unlinked ones. 

Your user story of "changes in base blocks rebuild of openshift" is impossible given the jobs are currently not linked. If they were, templating would be much harder / less standardized as we'd have to codify all the linkages.

File: buildchow/templates/core-jobs.yaml (hunk 3 above)
Comment (nickboldt) [inline buildchow/templates/core-jobs.yaml]: You mean you want to rebuild all the stable branch stuff using Beta2x bits?

Comment (maxandersen) [inline buildchow/templates/core-jobs.yaml]: I mean I would like to run these builds before actually making them live on the production jenkins instance. Wether that is beta2x or something else is secondary - but sure, beta2x works for me.

File: buildchow/templates/product-template.yaml (hunk 8 above)
Comment (maxandersen) [inline buildchow/templates/product-template.yaml]: You cannot seriously argue for not having actual releases used in our builds, can you ?

Why don't we just move back to using latest version from master ? 

Because then we can't actually trust our builds.

What in these releng scripts are *ever* expected to change ? if there are anything in them that requires changes between releases we have failed to actually make these reliable.

File: buildchow/templates/defaults.yaml [hunk 13]
```diff
@@ -4,35 +4,27 @@
...
+    jvm-memory: "-Xms512m -Xmx1536m"
     tests-folder: tests # for some it is all-tests
     component-responsible: ""
     scm-schedule: H H/8 * * * # every 8 hours starting a random hour. Removes need for each job to configure it (http://htmlpreview.github.io/?https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/hudson/triggers/TimerTrigger/help-spec.html)
-    download-cache:  "$JENKINS_HOME/static_build_env/jbds/download-cache"
-    maven-settings: "$JENKINS_HOME/../.m2/local-settings.xml" # need more reliable location
+    # these paths only exist on NFS-mounted slaves in the JBoss Jenkins cluster in mw.lab.eng.bos.redhat.com, which have a hudson user on them
+    # some slaves are configured with JENKINS_HOME = /qa/hudson_master/hudson_home/hudson_workspace, which is NOT the path we want here 
+    HUDSON_USER_HOME: "/home/hudson"
+    download-cache:  "{HUDSON_USER_HOME}/static_build_env/jbds/download-cache"
```
Comment (maxandersen) [inline buildchow/templates/defaults.yaml]: why is this HUDSON_USER_HOME still here ?

Just calling it something different does not make it more portable.

What is wrong with what was there before which was portable ?

File: buildchow/templates/core-jobs.yaml (hunk 2 above)
Comment (maxandersen) [inline buildchow/templates/core-jobs.yaml]: Just not following why this needs to be here when the component jobs don't need to refer to it at all.

And I do not follow why it would be hard to templetize with dependencies; this would just be a mention in the list of components what component they rely on. jjb should be possible to handle that pretty well IMO.

File: buildchow/templates/defaults.yaml (hunk 13 above)
Comment (nickboldt) [inline buildchow/templates/defaults.yaml]: /static_build_env only exists for some Jenkins slaves. Unless you plan to recreate it on your local Jenkins, it won't be there and won't work in $JENKINS_HOME.

Note too that based on https://engineering.redhat.com/rt/Ticket/Display.html?id=356526 not all slaves use /home/hudson so we should be using the ${HOME} folder of the user running Jenkins, which is what I set above. 

I still think this value should be externalized to the server config somehow.

Also not sure what to do about the reference to "local-settings.xml", which I don't see in our current jobs.

File: buildchow/templates/core-jobs.yaml (hunk 2 above)
Comment (nickboldt) [inline buildchow/templates/core-jobs.yaml]: In future we may have component jobs interconnected. For example... 

https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/jbosstools-javaee_master/ -> https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/jbosstools-javaee-tests-matrix_master/

So, the property makes sense, unless we decide we don't want tests to run without first rebuilding the plugins/features?

And yes, it's possible to handle the linkages. I've already set up a few, using two different Jenkins plugins (downstream-ext and trigger-parameterized-builds). It just means more deviations from the standard template.

File: buildchow/templates/defaults.yaml (hunk 13 above)
Comment (maxandersen) [inline buildchow/templates/defaults.yaml]: you are explicitly setting '/home/hudson' which the ticket you reference says will not exist so that does add up. If that dir does not exist $JENKINS_HOME is what should be used.

I *have* already tried these scripts, running a  local jenkins and I made it like this since it actually then works because teh one that creates `static_build_env/jbds/*` ensures the dir exists first. If you make it hardcoded to /home/hudson it will *not* work locally.

I do not follow how you think externalizing this value is better as hardcoded than just being relative to $JENKINS_HOME ? 

local-settings.xml is not relevant for the jobs is it ? That is a matter on what/how jenkins is configured. Separate from the templates.

File: buildchow/templates/core-jobs.yaml (hunk 2 above)
Comment (maxandersen) [inline buildchow/templates/core-jobs.yaml]: if we have some need in the future then handle it then. I honestly don't grok the work happening about adding a test matrix for javaee. Seems like an early experiment which is exactly the kind of things we could do better *outside* the production build servers via these templates. Didn't spot it being mentioned on mailing list or build.next - got any pointers ? 

The only deviation I see currently when setting up linkage is the list of dependencies that should be possible to encode as an argument to the template (or in worst case 2 templates)

File: buildchow/templates/defaults.yaml (hunk 13 above)
Comment (nickboldt) [inline buildchow/templates/defaults.yaml]: 1. JENKINS_HOME is not a defined variable on mv.lab.eng.bos slaves. Not sure where you found that. 

Recommend using either $HOME (which will be `/qa/services/hudson` or `/home/hudson` depending on the slave -- see references below - at item 4), or perhaps a combination of:

* $HUDSON_STATIC_ENV ($HOME/static_build_env) for the path to `static_build_env/jbds/` and 
* $HUDSON_HOME ($HOME/hudson_workspace) for the path to `.m2/settings.xml`. 

2. I'm not sure we even need `.m2/local-settings.xml`, which is why I commented it out -- you put that in, but I don't see a job config.xml which calls it. Maybe it's a leftover from an earlier time?

3. My point with setting HUDSON_USER_HOME: "/home/hudson" and wanting to externalize that into the server config, is that depending on the slave, you could set a different value. Thus for SOME Jenkins slaves it would be "/home/hudson" and for virt slaves it could be "/qa/services/hudson".

4. If you need examples of the differences in configs, see these:

http://jenkins.mw.lab.eng.bos.redhat.com/hudson/computer/vmg39-rhel6-x86_64/systemInfo
http://jenkins.mw.lab.eng.bos.redhat.com/hudson/computer/dspvmg03-rhel6-x86_64/systemInfo
vs.
http://jenkins.mw.lab.eng.bos.redhat.com/hudson/computer/dev128-virt2/systemInfo
http://jenkins.mw.lab.eng.bos.redhat.com/hudson/computer/dev125-virt2/systemInfo
http://jenkins.mw.lab.eng.bos.redhat.com/hudson/computer/dev127-virt1/systemInfo
http://jenkins.mw.lab.eng.bos.redhat.com/hudson/computer/dev126-virt1/systemInfo

5. The reason I chose HUDSON_USER_HOME instead of just HOME is that you might not be running Jenkins locally as your own user, but rather a different one; thus YOUR $HOME != $HUDSON_USER_HOME

Comment (nickboldt) [inline buildchow/templates/defaults.yaml]: Actually, that's not true. Just found it in 37 jobs. :(

    -gs /home/hudson/.m2/settings.xml

So... we should change that to HUDSON_USER_HOME, which we set to default value of ${{HOME}} if not set elsewhere

Comment (maxandersen) [inline buildchow/templates/defaults.yaml]: $JENKINS_HOME is set by default startup of jenkins. its used and available in jobs before so thats a shame. sending mail to ci-issues to clarify why they use HUDSON_HOME and not JENKINS_HOME.

There is something wrong if the shared hosts cannot agree on having an *environment variable* which one can use to derive this info.

Your code change moves this from an environment variable to a job template parameter - that just seems very wrong to me. This is bad separation of concerns.

Comment (maxandersen) [inline buildchow/templates/defaults.yaml]: btw. if we are introducing any variables we should refrain from using hudson - it is relying on jenkins...

I sent mail to ci-issues to clarify this because this is just not right if there is such big variation across the instances we can't rely on them.
